File: medication-tracker/backend/app/core/security_headers.py
from flask import Flask, Response

def init_security_headers(app: Flask) -> None:
    """Initialize security headers for the application"""
    
    @app.after_request
    def add_security_headers(response: Response) -> Response:
        # Enable HTTP Strict Transport Security (HSTS)
        response.headers['Strict-Transport-Security'] = 'max-age=31536000; includeSubDomains'
        
        # Prevent browsers from performing MIME type sniffing
        response.headers['X-Content-Type-Options'] = 'nosniff'
        
        # Enable Cross-Site Scripting (XSS) filter
        response.headers['X-XSS-Protection'] = '1; mode=block'
        
        # Prevent clickjacking
        response.headers['X-Frame-Options'] = 'SAMEORIGIN'
        
        # Content Security Policy
        csp = (
            "default-src 'self'; "
            "script-src 'self' 'unsafe-inline' 'unsafe-eval'; "
            "style-src 'self' 'unsafe-inline'; "
            "img-src 'self' data: https:; "
            "font-src 'self'; "
            "connect-src 'self'"
        )
        response.headers['Content-Security-Policy'] = csp
        
        # Referrer Policy
        response.headers['Referrer-Policy'] = 'strict-origin-when-cross-origin'
        
        # Permissions Policy
        response.headers['Permissions-Policy'] = (
            'camera=(), '
            'microphone=(), '
            'geolocation=()'
        )
        
        return response

File: medication-tracker/frontend/src/middleware/security-headers.js

const securityHeaders = {
  "Content-Security-Policy": {
    "default-src": [
      "'self'"
    ],
    "script-src": [
      "'self'",
      "'unsafe-inline'",
      "https://cdn.jsdelivr.net",
      "https://api.mixpanel.com"
    ],
    "style-src": [
      "'self'",
      "'unsafe-inline'",
      "https://fonts.googleapis.com"
    ],
    "img-src": [
      "'self'",
      "data:",
      "https:",
      "blob:"
    ],
    "font-src": [
      "'self'",
      "https://fonts.gstatic.com"
    ],
    "connect-src": [
      "'self'",
      "https://api.mixpanel.com",
      "https://your-api-domain.com",
      "wss://your-api-domain.com"
    ],
    "frame-src": [
      "'self'"
    ],
    "media-src": [
      "'self'"
    ],
    "object-src": [
      "'none'"
    ],
    "manifest-src": [
      "'self'"
    ],
    "worker-src": [
      "'self'",
      "blob:"
    ],
    "base-uri": [
      "'self'"
    ],
    "form-action": [
      "'self'"
    ],
    "frame-ancestors": [
      "'none'"
    ],
    "upgrade-insecure-requests": true
  },
  "Permissions-Policy": {
    "accelerometer": [],
    "camera": [],
    "geolocation": [],
    "gyroscope": [],
    "magnetometer": [],
    "microphone": [],
    "payment": [],
    "usb": []
  },
  "Strict-Transport-Security": "max-age=31536000; includeSubDomains; preload",
  "X-Content-Type-Options": "nosniff",
  "X-Frame-Options": "DENY",
  "X-XSS-Protection": "1; mode=block",
  "Referrer-Policy": "strict-origin-when-cross-origin"
};

module.exports = function applySecurityHeaders(app) {
  app.use((req, res, next) => {
    // Apply all security headers
    Object.entries(securityHeaders).forEach(([header, value]) => {
      if (typeof value === 'object') {
        if (header === 'Content-Security-Policy') {
          const cspValue = Object.entries(value)
            .map(([key, values]) => {
              if (key === 'upgrade-insecure-requests' && values === true) {
                return 'upgrade-insecure-requests';
              }
              if (Array.isArray(values)) {
                return `${key} ${values.join(' ')}`;
              }
              return '';
            })
            .filter(Boolean)
            .join('; ');
          res.setHeader(header, cspValue);
        } else if (header === 'Permissions-Policy') {
          const ppValue = Object.entries(value)
            .map(([key, values]) => {
              if (values.length === 0) {
                return `${key}=()`;
              }
              return `${key}=(${values.join(' ')})`;
            })
            .join(', ');
          res.setHeader(header, ppValue);
        }
      } else {
        res.setHeader(header, value);
      }
    });
    next();
  });
};
